test/distributed/org/apache/cassandra/distributed/test/ReadRepairEmptyRangeTombstonesTest.java

```diff
@@ -0,0 +1,323 @@
+import org.apache.cassandra.distributed.shared.AssertUtils;
+import org.apache.cassandra.service.reads.repair.ReadRepairStrategy;
+
+import static org.apache.cassandra.distributed.api.ConsistencyLevel.ALL;
+import static org.apache.cassandra.distributed.shared.AssertUtils.row;
+
+
+/**
+ * Test read repair when there are tombstone ranges, particularly to test that empty range tombstones are not emtited.
+ *
+ * See CASSANDRA-15924.
+ */
+@RunWith(Parameterized.class)
+public class ReadRepairEmptyRangeTombstonesTest extends TestBaseImpl
+{
+    private static final int NUM_NODES = 2;
+    private static final boolean[] BOOLEANS = new boolean[]{ false, true };
+
+    private static Cluster cluster;
+
+    /** The node to be used as coordinator */
+    @Parameterized.Parameter
+    public int coordinator;
+
+    /** The node to received the internal writes if no concrete node is specified */
+    @Parameterized.Parameter(1)
+    public int node;
+
+    /** Whether paging is used for the distributed queries */
+    @Parameterized.Parameter(2)
+    public boolean paging;
+
+    /** Whether the clustering order is reverse */
+    @Parameterized.Parameter(3)
+    public boolean reverse;
+
+    /** The read repair strategy to be used */
+    @Parameterized.Parameter(4)
+    public ReadRepairStrategy strategy;
+
+    @Parameterized.Parameters(name = "{index}: coordinator={0} node={1} paging={2} reverse={3} strategy={4}")
+    public static Collection<Object[]> data()
+    {
+        List<Object[]> result = new ArrayList<>();
+        for (int coordinator = 1; coordinator <= NUM_NODES; coordinator++)
+            for (int node = 1; node <= NUM_NODES; node++)
+                for (boolean paging : BOOLEANS)
+                    for (boolean reverse : BOOLEANS)
+                        for (ReadRepairStrategy strategy : ReadRepairStrategy.values())
+                            result.add(new Object[]{ coordinator, node, paging, reverse, strategy });
+        return result;
+    }
+
+    @BeforeClass
+    public static void setupCluster() throws IOException
+    {
+        cluster = init(Cluster.build(NUM_NODES).start());
+        cluster.schemaChange(withKeyspace("CREATE TYPE %s.udt (x int, y int)"));
+    }
+
+    @AfterClass
+    public static void teardownCluster()
+    {
+        if (cluster != null)
+            cluster.close();
+    }
+
+    /**
+     * Test range queries asking for an empty range with a tombstone range in one replica.
+     */
+    @Test
+    public void testEmptyRangeQueries()
+    {
+        tester().createTable("CREATE TABLE %s(k int, c int, PRIMARY KEY (k, c)) " +
+                             "WITH CLUSTERING ORDER BY (c %s) AND read_repair='%s'")
+                .executeInternal(1, "DELETE FROM %s USING TIMESTAMP 1 WHERE k=0 AND c>0 AND c<3")
+                .assertRowsDistributed("SELECT * FROM %s WHERE k=0 AND c>1 and c<=1")
+                .assertRowsDistributed("SELECT * FROM %s WHERE k=0 AND c>2 and c<=2")
```

Comment:
This is a nice translation and expansion of the old `emptyRangeTombstones1()`. One thing I'm rolling around a bit (and not sold on yet) is whether we should throw some quick assertions around the table-level read-repair metrics. I don't suppose we need yet another test to verify whether or not NONE works, but it could be a canary for other problems if the number of read-repairs ever diverges from expectations.

Reply:
I have added metrics checks for most of the queries here, and also in `ReadRepairQueryTypesTest`. `NONE` is checked in one of the JUnit params.